show scriptSig signature hash types in transaction decodes. fixes #3166

mruddy opened the pull request:
show scriptSig signature hash types. fixes #3166

Please give this a look and let me know if you'd like it changed or if I mis-understood the issue request.

The fix basically appends the scriptSig signature hash types, within parentheses, onto the end of the signature(s) in the various "asm" json outputs. That's just the first formatting idea that came to my mind.

Added some tests for this too.

laanwj commented:
Needs rebase.

laanwj commented on src/script/interpreter.cpp:
```diff
@@ -16,8 +16,6 @@
 
 using namespace std;
 
-typedef vector<unsigned char> valtype;
```
I don't like exposing `valtype` externally, the type name is too general.  Would prefer to change the function signatur to just
```c++
bool CheckSignatureEncoding(const vector<unsigned char> &vchSig, unsigned int flags, ScriptError* serror) {
```

mruddy commented:
Rebased, incorporated latest feedback about the function signature, and decided to actually replace the sighash byte on signatures when the decode occurs instead of just appending the decode after the sighash byte.

jtimon commented on src/script/script.cpp:
```diff
@@ -238,11 +240,11 @@ bool CScript::IsPushOnly() const
     return true;
 }
 
-std::string CScript::ToString() const
```
why remove std:: ?

mruddy replied:
Just for consistency within the file. That method was the only one using the prefix in that file because of the "using namespace std" up on line 12. That's all I was thinking when I did that.

jtimon replied:
We're doing the opposite in other places. That doesn't mean we should do it here. If you were touching the lines anyway I wouldn't have said anything, but increasing the diff only for consistency doesn't seem very interesting in my opinion. Anyway, feel free to ignore, just a small nit.

mruddy replied:
I agree with you, I'm going to put the prefixes back to reduce the diff in the next commit. Thanks for the info and for reviewing these changes.

jtimon commented on src/script/script.h:
```diff
@@ -11,12 +11,15 @@
 #include <string>
 #include <vector>
 
+#include <boost/assign/list_of.hpp>
```
Is this actually used in the .h ? If not, why not move it to the cpp or wherever it is used?

mruddy replied:
You might be right. I'll double check. I might have needed it at one point and then moved things around. If it's only needed in the cpp, I'll move it.

mruddy commented:
@jtimon your two feedback items are incorporated into this latest commit, thanks!

jtimon commented on src/script/script.cpp:
```diff
@@ -5,18 +5,11 @@
 
 #include "script.h"
 
+#include "script/interpreter.h"
```
Oh, I hadn't noticed the circular dependence...maybe an additional moveonly commit would solve that?
mhmm, moving CheckSignatureEncoding and the flags from interpreter to script would be an option. Maybe move the flags to consensus/consensus instead?
Another option is not having anything depending on the flags in script/script: CScript methods can be turned into functions taking a CScript as parameter.
Or maybe we can just let the circular dependency be for now.
Mhmm, I would like to know what other people think about this...I still don't even know what option I like more.

laanwj replied:
Yes, script should probably not depend on the interpreter.
I think we should move the disassembly logic out of the core/consensus data structure.

jtimon replied:
Would it be enough to move the opcode constants from script/interpreter to script/script in this PR to avoid this new include?
I haven't tried to compile that but I think that should be enough.

About IsPayToPublicKeyHash, do you mean ValueString shouldn't be a method of CScript ?
Otherwise, since CScript::ValueString uses IsPayToPublicKeyHash, it has to be in the same level or below, but not above.
If I have to suggest another place, I think I would go with script/standard, where other "static analysis" things are.

jgarzik commented:
Leaning towards closing.  Generally agree w/ the comment that this seems like a layering violation.

The intent is understandable, but it seems like the script show function is not the right place.

mruddy commented:
Yes, it should probably be closed. The more I worked on these changes, the less I liked them. I suppose I'll wait a day and close it unless someone objects.

jtimon commented:
I agree the intent is good and it can be done, just in script/standard rather than script/script.
If closing this PR and reopening another one that does the same in another file is clearer, let's do that.

jgarzik commented:
TxToUniv() dumps both script hex and script asm.

I wonder if an optional argument to script.ToString() would suffice?

Printing out the flags somewhere is certainly useful.  bitcoin-tx utility and RPC both call TxToUnix() when dumping the script verbosely.

mruddy commented:
I don't see how an optional argument to CScript::ToString would solve the layering concern.

That is caused by the script object needing to know if a piece of itself is a signature with a valid hash type. Right now, that logic is in the interpreter.

I spent some time looking at the layering.

I think IsValidSignatureEncoding could be moved over to script/bitcoinconsensus.cpp since it is entirely consensus-critical since BIP66.

Since the sig hash type is not part of consensus, only part of determining if a sig is standard, I think move IsDefinedHashtypeSignature and the sig hash types over to script/standard.cpp and .h respectively.

IsLowDERSignature is effectively dead code.

Then, CheckSignatureEncoding could also go to script/standard.cpp since it's not consensus and does not seem like it could only ever be used by the interpreter.

I think at that point the layering concern is covered.

Next, what lead to the klutziness in my code, that I wound up not really liking, was the need to answer the question of if the current script was a scriptSig. A lot of what I did to ValueString was to avoid mis-interpreting non-scriptSig data as a signature.

It seems like a flag could be set on the script object, or more object oriented, just use a SignatureScript type object when these objects are created (and PubKeyScript objects in other cases). Then this decode ticket would be an override to SignatureScript::ToString.

But, refactoring all that would obscure this change and is a lot of work for this seemingly low-value sighash decode. It'd turn into a train of refactoring tickets followed by the last ticket to actually do the simple decode. There's also probably a reason why all that refactoring would be misguided or error-prone that I don't foresee yet.

All my hacking to avoid refactoring is what makes me think this should just be closed.

Also, TxToUniv is only used by bitcoin-tx, CScript::ToString is the common dependency of RPC and bitcoin-tx.
The RPC dependencies go through, TxToJSON (decoderawtransaction and getrawtransaction) and ScriptPubKeyToJSON (decodescript).
That reason, and thinking that a script should know how to show/display/decode itself, is why I targeted CScript::ToString as the place to make these changes.

jtimon commented:
> I think IsValidSignatureEncoding could be moved over to script/bitcoinconsensus.cpp since it is entirely consensus-critical since BIP66.

It is in script/interpreter, so it's already part of libconsensus, see: https://github.com/bitcoin/bitcoin/blob/master/src/Makefile.am#L365

> Since the sig hash type is not part of consensus...
> Then, CheckSignatureEncoding could also go to script/standard.cpp since it's not consensus and does not seem like it could only ever be used by the interpreter.

The sighash types and CheckSignatureEncoding are definitely part of consensus (and are used by the interpreter), please, just leave them there.
The only thing you need to move is the new code for converting to string to script/standard and I think all layering concerns are gone.

mruddy commented:
> It is in script/interpreter, so it's already part of libconsensus, see: https://github.com/bitcoin/bitcoin/blob/master/src/Makefile.am#L365

That's true. I was just thinking that determining whether a signature is validly encoded is generic enough that it could be useful without including the interpreter. Personal preference, I guess.

> The sighash types and CheckSignatureEncoding are definitely part of consensus (and are used by the interpreter), please, just leave them there.

I was assuming that to begin with, but IsDefinedHashtypeSignature is only called when the flags include SCRIPT_VERIFY_STRICTENC. The comment for SCRIPT_VERIFY_STRICTENC even states, "not used or intended as a consensus rule". The best I found was that scripts with a bad sighash type would fail to eval (and thus be non-standard), but that if they got into valid blocks we'd still have to accept those blocks. I suppose I could try it and see, but that's what made me say that the sig hash type is not actually part of consensus. Maybe we're using the word consensus in two different ways here.

> The only thing you need to move is the new code for converting to string to script/standard and I think all layering concerns are gone.

Oh, you mean move ValueString and mapSigHashTypes over to script/standard and add the CScript as an argument to ValueString? Yes, that may make sense. And, if we move the sighash stuff from interpreter over to standard, then it'll all be in the same place.

sipa commented:
STRICTENC is indeed not technically consensus-critical, but it is part of
the interpreter (as we want it for evaluation of standardness), which
consensus-critical code depends on. I don't see how you can disconnect it
from the interpreter.

In general I prefer to have the basic data type definitions (script/script,
CScript) to have as little dependencies and functionality as possible, as
they get pulled into everything, including consensus code. Ideally, they
only provide representatiin, serialization, deserialization, and basic
construction. Something like ToString() is mostly for debugging purposes
IMHO, and should not be used for actual conversion to well-defined other
representation. See for example the core_io module for transaction/block
conversion code.

jtimon commented:
> Oh, you mean move ValueString and mapSigHashTypes over to script/standard and add the CScript as an argument to ValueString? Yes, that may make sense. 

Yes. I like this PR, but the code should just be in script/standard instead of script/script or script/interpreter, it's only that.

> And, if we move the sighash stuff from interpreter over to standard, then it'll all be in the same place.

No, the sighash stuff is consensus critical so you don't want it in script/standard because you don't want the rest of what's in there for libconsensus.
I wouldn't oppose to move the sighash flags, SignatureHash, TransactionSignatureChecker and MutableTransactionSignatureChecker to a different file (still required for consensus), but I don't think that's a priority and is certainly not needed for this PR (or libconsensus).

mruddy commented:
@jtimon OK, I just rebased and made the layering changes that I think we were talking about. I didn't squish them into one commit so that you could easily diff.

I did look at what @sipa was talking about. I see he added a FormatScript in core_write.cpp that I could probably build off of instead (never saw it before because it's only referenced by tests). He might be onto something. At quick glance, I think I might be able to make things simpler by going the route he suggested. I'll look into it more and may open an alternate pull request to this one. But at least the latest commit here kind of avoids the layering concern now, I think.

jgarzik commented:
Better!

jtimon commented:
Now it's actually worse because you're putting the whole script/standard inside libconsensus.
In fact, that's what is giving you errors on travis https://travis-ci.org/bitcoin/bitcoin/jobs/72684953#L2093
 because you haven't adapted makefile.am to include script/standard in libconsensus.

But you're very close, you can just move the rest of CScript::ToString to the new function in script/standard. That way you will be able to pass only a const CScript& as parameter (no additional vch param, which is nice) and you will be making libconsensus actually smaller instead of bigger. 
I'm not sure how many places CScript::ToString is called, but it is just a mechanical replace for ToString(const CScript&) [or whatever you want to call it].

Sorry if I haven't clearer earlier about CScript::ToString: if you want to use the new function in script/standard, it cannot be a method of CScript but an independent function instead (because script/script should not depend on neither script/interpreter nor script/standard).

jtimon commented:
To try to be clearer: if you need to "de-encapsulate" anything from CScript - by creating a new child class that uses the protected arguments, by creating new getters or even by making a private/protected public - to be able to move CScript::ToString up, I'm totally fine with that.
I'm happy with any solution that ends up with equal or less code in libconsensus (https://github.com/bitcoin/bitcoin/blob/master/src/Makefile.am#L365), anything else is secondary to me with rewards to this PR (it also makes it "less risky" [I don't think your changes are risky even violating the layering]).
My preference would be to simply rename ```CScript::ToString()``` to ```script/standard::ToString(const CScript&)``` - a trivial change using regular expressions or patience (the compiler won't let you get this little step wrong, anyway). But better is better, so my only requirement for a utACK is what I said before: equal or less code in libconsensus (and less is a plus).

mruddy commented:
@jtimon that's a lot of useful input, thanks! I took that and actually refactored these changes twice.

I put each of the two ways into the last two commits.

The first commit refactored into a couple of formatter classes. It worked, but the second/next idea seemed a little cleaner.

The second commit does away with the formatter class idea of the previous commit and just moves CScript::ToString out into a function in core_write.cpp.

I think this last commit converges your input with sipa's input. I think the last one loops in everyone's input, doesn't layer violate, and reduces code from the consensus library.

I purposefully didn't choose to sighhash decode some scriptSigs that were just going to be logged since those usages were tangential to what i was trying to actually update for this ticket.

Let me know what you think of the latest. Thanks!

mruddy commented:
... or not. I see the travis windows builds failed because primitives/transaction.cpp is still part of libconsensus and references the ScriptToAsmStr for ToString'ing CTxIn and CTxOut objects. Guess I'll have to move those around too.

jtimon commented:
You are welcomed. I'm fine with putting new things in core_io as @sipa suggested, just not in libconsensus. ```CTransaction::ToString``` now calls ```ScriptToAsmStr``` so instead of putting script/standard in libconsensus, now you're putting core_io instead. The important thing is this: independently of where you chose to put the new code outside of libconsensus, you cannot include that new code from libconsensus. That's why I keep posting a link to the list of files that are part of libconsensus: https://github.com/bitcoin/bitcoin/blob/master/src/Makefile.am#L365
If you include more things in any of those files, travis will complain (unless you adapt the makefile properly to include the new file, in which case we will complain).

So, either don't adapt CTransaction::ToScript to the new code (seems the simpler solution) or also take CTransaction::ToString out of consensus (seems better but also more work).

jtimon commented:
Another possibility is to leave the new code in libconsensus like you had at the beginning (not my preference), but it still couldn't be in script/script since you need script/interpreter and that would cause a circular dependency. So even leaving the new code in libconsensus, you always need to rename CScript::ToString and move it our of script/script.

mruddy commented:
It turns out that resolving that last dependency was pretty easy.
I simply stopped using that script assembly formatting in the CTxIn and CTxOut ToString methods.
Those ToString's were only called by very low value debug and log print statements and the formatted scripts were being truncated immediately too. So, converting to hex instead in those cases isn't losing much, if anything at all, in my opinion. Doing it this way allowed the removal of the original CScript::ToString from libconsensus.

jtimon commented on src/primitives/transaction.cpp:
```diff
@@ -5,6 +5,7 @@
 
 #include "primitives/transaction.h"
 
+#include "core_io.h"
```
This is not necessary anymore, right?

jtimon commented on src/bitcoin-tx.cpp:
```diff
@@ -12,6 +12,7 @@
 #include "primitives/transaction.h"
 #include "script/script.h"
 #include "script/sign.h"
+#include "script/standard.h"
```
This is core_io now.

jtimon commented on src/test/script_tests.cpp:
```diff
@@ -9,6 +9,7 @@
 #include "key.h"
 #include "keystore.h"
 #include "script/script.h"
+#include "script/standard.h"
```
This is core_io now.

jtimon commented:
This looks like a reasonable solution.
Apart from the latest little nits, consensus-safe-review ACK (I also slightly reviewed the new functions, but didn't look much at the tests).

mruddy commented:
@jtimon You are correct about those latest nits. I made the updates and squashed+rebased everything back down to the one latest commit. I also updated the release notes description at the same time. Should be all good now. Thanks for the help, it was good working with you on this!

jtimon commented:
Thanks to you for your patience when improving this. Everything looks good, re-utACK.

sipa commented on src/core_write.cpp:
```diff
@@ -54,6 +55,67 @@ string FormatScript(const CScript& script)
+            if (vch.size() <= static_cast<vector<unsigned char>::size_type>(4)) {
+                str += strprintf("%d", CScriptNum(vch, false).getint());
+            } else {
+                if (fAttemptSighashDecode) {
```
Nit: the !script.IsUnspendable() check can move here, I believe.

sipa commented:
Untested ACK, but the tests look convincing.

mruddy committed:
Resolve issue 3166.
These changes decode valid SIGHASH types on signatures in assembly (asm) representations of scriptSig scripts.
This squashed commit incorporates substantial helpful feedback from jtimon, laanwj, and sipa.

mruddy commented:
@sipa nit addressed. I also added a new test case just to cover a specific low probability case related to the nit to make sure it was covered.

jgarzik commented:
ut ACK - looks ready to merge

arielgabizon commented on src/primitives/transaction.cpp:
```diff
@@ -36,7 +36,7 @@ std::string CTxIn::ToString() const
         str += strprintf(", coinbase %s", HexStr(scriptSig));
     else
-        str += strprintf(", scriptSig=%s", scriptSig.ToString().substr(0,24));
+        str += strprintf(", scriptSig=%s", HexStr(scriptSig).substr(0, 24));
```
was there a reason `HexStr` wasn't used here before?

mruddy replied:
It was just more verbose (someone's personal preference, I guess), see what was removed `CScript::ToString`: https://github.com/bitcoin/bitcoin/commit/af3208bfa6967d6b35aecf0ba35d9d6bf0f8317e#diff-f7ca24fb80ddba0f291cb66344ca6fcb
